components/mpas-ocean/src/shared/mpas_ocn_frazil_forcing.F

```diff
@@ -669,6 +756,297 @@ subroutine ocn_frazil_forcing_build_arrays(domain, meshPool, forcingPool, stateP
+
+
+        ! initialize the sum of new frazil ice created
+        sumNewFrazilIceThickness = 0.0_RKIND
+        sumNewThicknessWeightedEcosysContent(:) = 0.0_RKIND
+
+        ! loop from maximum depth of frazil creation to surface
+        do k = kBottomFrazil, minLevelCell(iCell), -1
+
+          ! get freezing temperature
+          oceanFreezingTemperature = ocn_freezing_temperature(salinity=activeTracers(indexSalinity,k,iCell), &
+                                                              pressure=pressure(k,iCell), &
+                                                              inLandIceCavity=underLandIce)
+
+          potential = layerThickness(k,iCell) * config_specific_heat_sea_water &
+                    * rho_sw * (activeTracers(indexTemperature,k,iCell) - oceanFreezingTemperature)
+
+          freezingEnergy = max(0.0_RKIND, -potential)
+          meltingEnergy = max(0.0_RKIND, potential)
+
+          if (freezingEnergy > 0) then
+
+             ! new frazil ice formation measured in meters
+             newFrazilIceThickness = freezingEnergy / (config_frazil_heat_of_fusion * config_frazil_ice_density)
+
+             ! limit the frazil formed appropriately
+             newFrazilIceThickness = min(newFrazilIceThickness, layerThickness(k,iCell) * config_frazil_fractional_thickness_limit)
+
+             ! keep track of sum of frazil ice
+             sumNewFrazilIceThickness = sumNewFrazilIceThickness + newFrazilIceThickness
+
+             do n = 1, nTracersEcosys
+                frazilEcosys = (config_frazil_ice_porosity*usePorosity + (1-usePorosity)) * ecosysTracers(n, k, iCell)
+                ! needs to be an array
+                newThicknessWeightedEcosysContent = newFrazilIceThickness * frazilEcosys
+
+                ! bgc is extracted with the frazil
+                ! forcingPool
+                frazilEcosysTracersTendency(n,k,iCell) =  -newThicknessWeightedEcosysContent / dt
+
+                sumNewThicknessWeightedEcosysContent(n) = sumNewThicknessWeightedEcosysContent(n) + &
+                    newThicknessWeightedEcosysContent
+             end do
+
+          else
+            ! ocean water is warm enough to melt frazil
+
+            ! test to see if there is frazil to be melted
+            if (sumNewFrazilIceThickness > 0.0_RKIND) then
+
+               ! Frazil melting
+               meltedFrazilIceThickness = meltingEnergy / (config_frazil_heat_of_fusion * config_frazil_ice_density)
+
+               ! limit melting by what there is to melt
+               meltedFrazilIceThickness = min(meltedFrazilIceThickness, sumNewFrazilIceThickness)
+
+               ! limit melting by fraction of layer thickness
+               meltedFrazilIceThickness = min(meltedFrazilIceThickness, layerThickness(k,iCell) &
+                                       * config_frazil_fractional_thickness_limit)
+
+               do n = 1, nTracersEcosys
+
+                  ! assign some ecosys tracer to the melted ice
```

Comment:
It might be worth pointing out here and/or in the PR description that ecosystem tracers are released from frazil ice only if there was a frazil tendency in that location in that time step